### PR TITLE
feat(merge): native hunk-level merge engine (heddle#79)

### LINKED ISSUE #79: merge: native hunk-level merge engine (replaces whole-file conflicts when semantic declines)
## Split from [heddle#68](https://github.com/HeddleCo/heddle/issues/68) AC #3

**This is the load-bearing piece for heddle git-overlay daily use.** Per the heddle#54 rebase trip report, semantic merge bails to whole-file conflicts on structural changes, and git rebase produced 3 small hunks vs heddle's 7 whole-file collisions for the same task.

## Strategic framing

Heddle is meant to eventually replace git. The fix is **not** to fall back to \`git merge-file\` — that bakes in git dependence. The fix is to implement heddle's own line-based 3-way merge engine matching git's baseline competence, with the semantic resolver layered on top.

## Layered architecture target

- **Semantic merge succeeds** → auto-resolved, no markers
- **Semantic merge declines** → fall through to **heddle-native** \`text_hunk_merge\` for that file. Hunk conflicts emitted in canonical marker format (depends on heddle#80 marker fix landing)
- **Hunk-level merge sees no conflicts** (disjoint line ranges) → auto-resolved at line level
- **Hunk-level merge sees conflicts** → emit canonical hunk markers, user resolves

## Acceptance criteria

- [ ] Implement \`heddle::merge::text_hunk_merge\` — 3-way line-based merge with conflict detection. Same algorithm git uses (recursive longest-common-subsequence on text, conflict detection on overlapping hunks).
- [ ] Wire it into \`merge --semantic\`: when the semantic resolver returns "decline" for a path, run hunk_merge on that path instead of emitting a whole-file conflict.
- [ ] **Red commit**: fixture pair where semantic declines but line ranges are disjoint → assert auto-resolution at line level (no markers in the output).
- [ ] **Red commit**: fixture pair where semantic declines AND line ranges overlap → assert hunk-level markers (not whole-file).
- [ ] Bench: confirm the layered approach (semantic try → hunk-level fallback) doesn't add measurable latency on the no-conflict path.

## Public API

If \`text_hunk_merge\` exposes a useful primitive (\`MergeOutcome\` enum), surface it as public API in \`heddle-merge\` (or wherever the merge crate lives) so other heddle tooling can compose on it.

## Suggested effort

**max** or **xhigh** — this is the load-bearing engineering. The 3-way line-based merge with all its edge cases (rename detection? line-ending normalization? trailing-newline semantics? whitespace-only changes?) is real work. Reading git's merge algorithm source for reference is recommended.

## Blocked by

None directly, but landing alongside or after the marker-fix issue (heddle#80) means the emitted hunk markers are well-formed.

## Definition of Done
- [ ] **Red-commit first**: failing fixtures pushed before implementation.
- [ ] PR description includes the merge-fixture test matrix + bench numbers.
- [ ] Cross-repo PRs linked back to this issue.
- [ ] Acceptance criteria all checked.

### PR DESCRIPTION
**Pull request by thethorne48**
## Summary

- New `heddle_semantic::merge::text_hunk_merge` — a diff3-style line-level three-way merger with per-hunk conflict markers. Built on `similar::capture_diff_slices` (Histogram). Closes the heddle#54 trip-report failure mode where heddle's prior single-range merger bailed to whole-file conflicts on any multi-hunk diff that git rebased into 3 small hunks.
- Wired into both `heddle merge` (`merge_algo::three_way_content_merge`) and `heddle rebase` (`rebase_ops::auto_merge_text_lines`) — Rule-7 sweep caught the duplicate single-range algorithm in rebase, which heddle#54 was the trip report for. Both now route through the new engine when the `semantic` feature is on (default).
- Markers emitted are well-formed at column 0 (heddle#78-compatible), and the fallback whole-file paths (Binary, modify/delete, tree-entry mismatches) keep using the just-fixed `format_conflict_content` emitter — no duplicate marker logic.

Pipeline position now matches the brief's layered model:

```
semantic merge succeeds → resolved (existing path)
semantic declines       → text_hunk_merge (NEW)
  hunks disjoint        → auto-resolved at line level
  hunks overlap         → per-hunk markers (not whole-file)
  binary / del-vs-mod   → fall back to existing whole-file emitter
```

Algorithm: walk `base` once, identify *stable* lines (lines aligning to the current expected position in BOTH ours and theirs) vs *unstable hunks* (everything else), classify each hunk one of four ways (ours==base → take theirs / theirs==base → take ours / ours==theirs → take either / else → conflict markers). Cost: O(n) walk over base on top of two LCS computations.

Public API in `heddle_semantic`:

```rust
pub fn text_hunk_merge(base: &[u8], ours: &[u8], theirs: &[u8]) -> MergeOutcome;
pub fn text_hunk_merge_with_markers(
    base: &[u8], ours: &[u8], theirs: &[u8], markers: ConflictMarkers<'_>,
) -> MergeOutcome;

pub enum MergeOutcome {
    Clean(Vec<u8>),
    Conflicts { merged_bytes_with_markers: Vec<u8>, conflict_count: usize },
    Binary,
    DeleteVsModify,
}
```

## Algorithm + crate choice

Built on `similar` (already a workspace dep via `objects::worktree_diff`). `similar::capture_diff_slices(Algorithm::Histogram, base_lines, other_lines)` gives Equal-op alignments — the only LCS primitive needed. The diff3 walk on top is ~80 lines and gives byte-exact control over markers (critical given heddle#78's marker validator). Chose to implement the walk myself rather than adopt `diffy::merge` to: (a) avoid adding a new dep, (b) keep marker emission inside the crate so heddle's well-formedness invariants are local, (c) preserve heddle's parallel-thread append UX (concat-at-same-anchor — see "Edge cases" below).

## Red-commit DoD

Red: `47907fd test(merge): red — text_hunk_merge multi-hunk fixtures (heddle#79)` — stub `text_hunk_merge` returned a single whole-file conflict regardless of input; 8/10 new unit tests failed against it. Green: `084a3de feat(merge): green — diff3 hunk-level engine in heddle-semantic (heddle#79)` — 10/10 green, 12/12 after the same-anchor-concat UX tests, full crate 106/106.

Three brief-mandated fixtures all pass:

1. `disjoint_multi_hunks_auto_resolve` — ours edits two disjoint hunks, theirs a third → Clean.
2. `overlapping_hunks_produce_hunk_markers` — both edit same line → markers + surrounding lines preserved verbatim.
3. `large_file_disjoint_3line_changes_resolve_cleanly` — heddle#54 shape: 185-line file, 3-line edits each side at distant positions → Clean.

## Rule-7 findings

| Finding | Location | Action |
|---|---|---|
| Duplicate single-hunk algorithm in rebase replay | `crates/cli/src/cli/commands/rebase/rebase_ops.rs::auto_merge_text_lines` (90 lines) | **Replaced** in this PR — 5-line call to `text_hunk_merge`. This is the actual code heddle#54 measured. |
| `format_conflict_content` callers (5 sites) | `executor.rs` whole-file / modify-delete / tree-entry-mismatch paths | **Kept** — these handle cases `text_hunk_merge` can't represent (no base text, file-vs-dir, etc.). Marker format already fixed by heddle#78 (#82, merged). No duplication: my new emitter and the existing one both produce column-0 markers. |
| Thread refresh merge | `crates/cli/src/cli/commands/merge/mod.rs::try_three_way_merge_between_tips` | **No change needed** — converges on `three_way_content_merge`, picks up the new engine automatically. |
| `crates/repo/src/repository_provenance/merge.rs` | tree-level only, no content merge | None — no whole-file content emission. |

## Edge cases

| Case | Treatment |
|---|---|
| Trailing newline divergence (`\n` vs no-`\n`) | Preserved byte-exact; not itself a conflict trigger. Covered by `trailing_newline_divergence_does_not_conflict`. |
| CRLF vs LF | Preserved verbatim; no normalization. Documented in `merge/mod.rs`. |
| Whitespace-only trailing edits | If both sides differ only in trailing whitespace, pick cleaner. Lives in `whitespace.rs`. |
| Binary files | NUL scan over first 8 KiB. Returns `Binary` → caller falls back to whole-file conflict (matches git). |
| One side deletes, other modifies | Detected at the tree layer; reserved variant `DeleteVsModify` (never produced by `text_hunk_merge` itself). |
| Same-anchor insertions, different content | **Preserved heddle UX**: concat ours-then-theirs rather than conflict. Validated by `test_merge_auto_merges_non_overlapping_same_file_appends_from_threads` + two new unit tests. |
| **Deferred**: concurrent file rename + edit | Per brief — file as follow-up. |

## Bench

`merge::tests::bench_no_conflict_disjoint_hunks_large_file` exercises the heddle#54-shape worst case (1000 lines, 3 disjoint hunks per side):

| Build | per-call | iterations |
|---|---|---|
| `--release` | ~10 ms | 50 in 497 ms |
| `dev` | ~155 ms | 50 in 7.7 s |

Well under any user-perceptible threshold for a real-time merge. Budgets in the test catch pathological regressions.

## File layout

```
crates/semantic/src/merge/
  mod.rs         (101 L)  public API
  diff3.rs       (172 L)  algorithm: run + walk + classify
  lines.rs       ( 41 L)  split_lines + build_alignment
  markers.rs     ( 60 L)  ConflictMarkers + emit_conflict + emit_lines
  whitespace.rs  ( 61 L)  trailing-ws-equivalence helpers
  preflight.rs   ( 13 L)  binary detection
  tests.rs       (--)     fixtures
```

All files under the 300-line cap.

Closes HeddleCo/heddle#79

## Test plan
- [x] 12/12 new `merge::*` unit tests green (3 brief-mandated + 9 edge cases + 1 bench)
- [x] 106/106 `heddle-semantic` lib tests green
- [x] 291/291 `heddle-cli` lib tests green
- [x] 36/36 `heddle-cli` state-management/merge integration tests green
- [x] 3/3 state-management/rebase tests green (including the disjoint-edits replay test)
- [x] formal-spec `merge_safety_invariants` green
- [x] `cargo clippy -p heddle-semantic -p heddle-cli --all-targets -- -D warnings`: clean
- [x] Files under 300-line cap; functions under 100-line cap

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---
<a href="https://app.blacksmith.sh/HeddleCo/codesmith/heddle/pr/84"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-light.svg">View in Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need.</sup>

- [ ] Let Codesmith autofix CI failures and bot reviews

### CODE REVIEW

File: crates/semantic/src/merge/diff3.rs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+    their_slice: &[&[u8]],
+    markers: ConflictMarkers<'_>,
+) {
+    if slice_eq(our_slice, base_slice) {
+        emit_lines(out, their_slice);
+    } else if slice_eq(their_slice, base_slice) {
+        emit_lines(out, our_slice);
+    } else if slice_eq(our_slice, their_slice) || trailing_ws_equal(our_slice, their_slice) {
+        emit_lines(out, prefer_clean(our_slice, their_slice));
+    } else if base_slice.is_empty() && !our_slice.is_empty() && !their_slice.is_empty() {
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/semantic/src/merge/diff3.rs]: Fixed in 14264d5: added a `compose_disjoint` pass in `crates/merge/src/diff3.rs` that runs after the four-way classifier. It classifies each side's LCS diff against `base` into per-line actions (Keep/Replace/Delete) plus per-gap insertions, then emits the composition iff the two patches commute (no overlap at any base line). Both side orderings (`ours=insert/theirs=edit` and the swap) are tried; any overlap bails to the existing conflict path. New fixtures: `insertion_plus_adjacent_edit_resolves_cleanly`, `insertion_plus_adjacent_edit_symmetric`, `insertion_plus_adjacent_delete_resolves_cleanly`, `insertion_at_eof_plus_edit_resolves_cleanly`, `insertion_at_bof_plus_edit_resolves_cleanly`, plus a `both_sides_modify_same_line_still_conflicts` negative case.

File: crates/cli/src/cli/commands/rebase/rebase_ops.rs [hunk 2]
```diff
@@ -315,114 +315,36 @@ fn try_auto_merge_textual_change(
...
+#[cfg(not(feature = "semantic"))]
+fn auto_merge_text_lines(
+    _base: &[u8],
+    _current: &[u8],
+    _incoming: &[u8],
+) -> Result<Option<Vec<u8>>> {
+    // No semantic feature → no native hunk-level merge engine. Decline
+    // auto-merge; caller falls back to `blob_contains_both` or marks
+    // the file conflicted.
+    Ok(None)
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/cli/src/cli/commands/rebase/rebase_ops.rs]: Fixed in 14264d5: extracted the text hunk merger into a new always-on `heddle-merge` crate (`crates/merge/`). `auto_merge_text_lines` no longer has the `#[cfg(not(feature = "semantic"))]` branch that unconditionally returned `Ok(None)` — it routes through `merge::text_hunk_merge` regardless of the `semantic` feature, so a `--no-default-features` rebase keeps the prior textual auto-merge behavior the brief describes. Verified by building and testing with `cargo test -p heddle-cli --no-default-features --features git-overlay,native,local` (266 lib tests green).

File: crates/cli/src/cli/commands/merge/merge_algo/executor.rs [hunk 3]
```diff
@@ -228,125 +228,75 @@ fn three_way_content_merge(
...
+}
 
-    text.split_inclusive('\n')
-        .map(ToString::to_string)
-        .collect()
+fn load_blob_content(store: &dyn ObjectStore, hash: &ContentHash) -> Result<Vec<u8>> {
+    Ok(store
+        .get_blob(hash)?
+        .map(|blob| blob.content().to_vec())
+        .unwrap_or_default())
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/cli/src/cli/commands/merge/merge_algo/executor.rs]: Fixed in 14264d5: `load_blob_content` now returns an anyhow error when `get_blob(hash)?` is `None`, instead of `unwrap_or_default()`-ing into empty bytes. The error names the missing hash and the affected path so an operator can locate the corrupt entry. Routed `content_conflict_merge` and `modify_delete_conflict_merge` through the same helper as a Rule-7 sweep — they had the same silent-corruption shape (a missing side would appear as an empty half of the conflict). Two new unit tests cover the fail-fast and success paths via `InMemoryStore`: `missing_merge_blob_surfaces_as_error_not_silent_empty` and `present_merge_blob_loads_content`.

File: crates/cli/src/cli/commands/merge/merge_algo/executor.rs [hunk 4]
```diff
@@ -228,125 +228,75 @@ fn three_way_content_merge(
...
+    their_hash: &ContentHash,
+    path: &str,
+    conflicts: &mut Vec<String>,
+    labels: ConflictLabels<'_>,
+) -> Result<ContentHash> {
+    // No semantic feature → no native hunk-level merge engine. Fall back
+    // to whole-file conflict markers (preserves prior behavior for users
+    // who explicitly built with `--no-default-features` without
+    // re-enabling `semantic`).
+    content_conflict_merge(store, our_hash, their_hash, path, conflicts, labels)
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/cli/src/cli/commands/merge/merge_algo/executor.rs]: Fixed in 14264d5 by the same architectural move as the rebase finding: the text hunk merger now lives in its own always-on `heddle-merge` crate. `text_hunk_merge_blobs` is no longer gated on `#[cfg(feature = "semantic")]`, so non-semantic merge builds route every divergent text merge through `text_hunk_merge_with_markers` instead of dropping straight into `content_conflict_merge`. Previously-auto-resolvable cases stay auto-resolved on `--no-default-features` CLI builds.

File: crates/semantic/src/merge/whitespace.rs [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+/// True if `a` and `b` would be equal after stripping trailing whitespace
+/// (` `, `\t`) from each line (line endings preserved or absent equally).
+pub(super) fn trailing_ws_equal(a: &[&[u8]], b: &[&[u8]]) -> bool {
+    if a.len() != b.len() {
+        return false;
+    }
+    a.iter()
+        .zip(b.iter())
+        .all(|(x, y)| strip_trailing_ws(x) == strip_trailing_ws(y))
```
[the remark this answers is not in the document]
Comment (thethorne48) [reply crates/semantic/src/merge/whitespace.rs]: Fixed in 14264d5: replaced `strip_trailing_ws` with `trailing_ws_key`, a comparison key that preserves the line-ending bytes (`\r\n` / `\n` / `""`) verbatim alongside the body-with-trailing-spaces-stripped slice. Only intra-line trailing spaces and tabs are stripped now; line endings are no longer folded out, so CRLF vs LF on the same edit produces a real conflict instead of being auto-resolved via `prefer_clean`. New fixture `crlf_vs_lf_on_same_line_conflicts` exercises the path; `trailing_space_only_difference_still_auto_resolves_with_same_endings` keeps a green case for the legitimate trailing-whitespace folding.

**Comment by thethorne48** [comment]
@codex please re-review — pushed `14264d5` addressing all 5 findings from r1.

- **P1 (insertion+edit hunk split)** — added `compose_disjoint` pass to `crates/merge/src/diff3.rs`; classifies each side's LCS diff into per-line actions + per-gap insertions and emits the composition when patches commute. 6 new fixtures including the `X NEW Y / X Y'` shape and a negative case.
- **P2 (non-semantic regressions in both executor.rs and rebase_ops.rs)** — extracted the text hunk merger into a new always-on `heddle-merge` crate (`crates/merge/`). Dropped the `#[cfg(feature = "semantic")]` gates around `text_hunk_merge_blobs` and `auto_merge_text_lines`. `--no-default-features` builds now keep textual auto-merge as before.
- **P2 (missing-blob silent corruption)** — `load_blob_content` now returns an anyhow error naming hash + path instead of `unwrap_or_default()`. Routed `content_conflict_merge` + `modify_delete_conflict_merge` through the same helper (Rule-7 sweep).
- **P2 (CRLF vs LF in `trailing_ws_equal`)** — replaced strip-and-compare with a key that preserves line-ending bytes; only intra-line trailing whitespace is folded out.

Tests: heddle-merge 21 (was 10) · heddle-cli +2 unit · workspace all green · bench within budget.
